Add option to use mempool on OOM

dsjohns2 opened the pull request:
MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a use_on_oom bool with each MemPool.

Usage:
Users can optionally specify a ``use_on_oom`` bool (which is False by default) during MemPool creation. If true, then the CUDACachingAllocator will be able to use memory in this pool as a last resort instead of OOMing.

```
pool = torch.cuda.MemPool(allocator, use_on_oom=True)
with torch.cuda.use_mem_pool(pool):
    a = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
del a
# at the memory limit, this will succeed by using pool's memory in order to avoid the oom
b = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
```

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151487



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

dsjohns2 committed:
use mempool on oom

[ghstack-poisoned]

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 reviewed (commented):
Code looks good to me. I will defer to @syed-ahmed @eqy @zdevito @ngimel for final decision.

eqy reviewed (commented):
No objections from me, will defer to @syed-ahmed 

One minor question: what is the existing behavior today? Would `empty_cache` or its equivalent be done on the mempools today, and if not, would that be an intermediate alternative?

eqy commented on test/test_cuda.py:
```diff
@@ -4884,6 +4884,27 @@ def test_no_triton_on_import(self):
 
 @unittest.skipIf(not TEST_CUDA, "CUDA not available, skipping tests")
 class TestMemPool(TestCase):
+    def _setup_mempool_limited_memory_test(self, additional_allowed_memory_in_mb):
```
I'm wondering if this needs a `serialTest` decorator as exactly tracking allocation stats seems scary if we allow parallel execution of tests

eqy commented on test/test_cuda.py:
```diff
@@ -4884,6 +4884,27 @@ def test_no_triton_on_import(self):
 @unittest.skipIf(not TEST_CUDA, "CUDA not available, skipping tests")
 class TestMemPool(TestCase):
+    def _setup_mempool_limited_memory_test(self, additional_allowed_memory_in_mb):
+        os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
```
Is this the preferred way that `expandable_segments` should be set? Should this be coalesced with other expandable segments tests e.g., in `test_cuda_expandable_segments.py` to avoid polluting tests that don't expect this behavior?

dsjohns2 replied:
Thanks for the comments! Makes sense. 

We shouldn't actually need expandable_segments to test this. Is the suggestion to add this test to the run_tests in test_cuda_expandable_segments.py? My understanding is that just reruns a bunch of tests with expandable segments turned on.

eqy replied:
If it's not needed I would just remove the conf setting---it's difficult to maintain and every `setenv/getenv` is a potential thread-safety conflict point as well.

eqy commented on test/test_cuda.py:
```diff
@@ -4884,6 +4884,27 @@ def test_no_triton_on_import(self):
+        return device, dtype
+
+    def _teardown_mempool_limited_memory_test(self):
+        os.environ["PYTORCH_CUDA_ALLOC_CONF"] = ""
```
See above, I'm not sure we expect `PYTORCH_CUDA_ALLOC_CONF` to be re-parsed every time? https://github.com/pytorch/pytorch/blob/81c4369d813facf39313dfd481adc71704cbc2c1/c10/cuda/CUDAAllocatorConfig.h#L81

eqy commented on test/test_cuda.py:
```diff
@@ -4884,6 +4884,27 @@ def test_no_triton_on_import(self):
+    def _teardown_mempool_limited_memory_test(self):
+        os.environ["PYTORCH_CUDA_ALLOC_CONF"] = ""
+        torch.cuda.memory.empty_cache()
+        torch.cuda.memory.set_per_process_memory_fraction(1.0)
```
Might need to stash the previous value instead of assuming it's 1.0 https://github.com/pytorch/pytorch/blob/81c4369d813facf39313dfd481adc71704cbc2c1/torch/cuda/memory.py#L194

eqy commented on test/test_cuda.py:
```diff
@@ -5000,6 +5021,96 @@ def test_mempool_with_allocator(self):
+        # default pool [] 0 mb
+
+        # expect that we used same memory address for both a and c
+        self.assertEqual(b_dataptr, c_dataptr)
```
For completeness, should there be a check that the "fallback" allocation can be properly recycled once C is deleted?

syed-ahmed replied:
+1

dsjohns2 replied:
Good point, added this check, thanks!

eqy reviewed (approved):
Left a few nits but no objections from me, will defer to @syed-ahmed 

One minor question: what is the existing behavior today? Would `empty_cache` or its equivalent be done on the mempools today, and if not, would that be an intermediate alternative?

syed-ahmed commented on torch/cuda/memory.py:
```diff
@@ -1132,8 +1132,8 @@ class MemPool(_MemPool):
 
-    def __init__(self, allocator: Optional[_cuda_CUDAAllocator] = None):
-        super().__init__(allocator, True)
+    def __init__(self, allocator: Optional[_cuda_CUDAAllocator] = None, useOnOOM=False):
+        super().__init__(allocator, True, useOnOOM)
```
+1

syed-ahmed commented on test/test_cuda.py:
```diff
@@ -5000,6 +5021,96 @@ def test_mempool_with_allocator(self):
+
+        # pool's destructor calls emptyCache()
+        del pool_use, pool_do_not_use
+
```
Thanks for the test! Looks great!

syed-ahmed commented on torch/cuda/memory.py:
```diff
@@ -1132,8 +1132,8 @@ class MemPool(_MemPool):
 
```
Please add documentation here on what `use_on_oom` is supposed to do.

dsjohns2 replied:
Added documentation here and to cuda.rst

syed-ahmed reviewed (approved):
Interesting feature! Overall looks good to me. If you could leave a note [in this section ](https://github.com/pytorch/pytorch/blob/6ced5e68401471dab9decb2b8f8f61478a5db1cc/docs/source/notes/cuda.rst#mixing-different-cuda-system-allocators-in-the-same-program) summarizing the use case and a small snippet on how a user can enable this feature, I would really appreciate it.

ngimel reviewed (approved):
lgtm, module renaming to `use_on_oom`

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a useOnOOM bool with each MemPool.

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 commented:
> One minor question: what is the existing behavior today? Would `empty_cache` or its equivalent be done on the mempools today, and if not, would that be an intermediate alternative?

From [documentation](https://github.com/pytorch/pytorch/blob/6ced5e68401471dab9decb2b8f8f61478a5db1cc/docs/source/notes/cuda.rst#mixing-different-cuda-system-allocators-in-the-same-program): "To reclaim memory, users will first need to ensure nothing is using the pool. When none of the tensors are holding a reference to the pool, torch.cuda.empty_cache will be called internally on deletion of the pool, hence returning all the memory to the system." 

So existing behavior is like `pool_do_not_use` in test case where it will hold onto memory until pool deletion.

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a use_on_oom bool with each MemPool.

Usage:
Users can optionally specify a ``use_on_oom`` bool (which is False by default) during MemPool creation. If true, then the CUDACachingAllocator will be able to use memory in this pool as a last resort instead of OOMing.

```
pool = torch.cuda.MemPool(allocator, use_on_oom=True)
with torch.cuda.use_mem_pool(pool):
    a = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
del a
# at the memory limit, this will succeed by using pool's memory in order to avoid the oom
b = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
```

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 commented:
@pytorchbot merge

dsjohns2 commented:
It seems like touching MemPool.cpp is causing linter to have an issue due to https://github.com/pytorch/pytorch/pull/148948.

dsjohns2 commented:
I think we can just add 
```
// NOLINTNEXTLINE(misc-use-internal-linkage)
```

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a use_on_oom bool with each MemPool.

Usage:
Users can optionally specify a ``use_on_oom`` bool (which is False by default) during MemPool creation. If true, then the CUDACachingAllocator will be able to use memory in this pool as a last resort instead of OOMing.

```
pool = torch.cuda.MemPool(allocator, use_on_oom=True)
with torch.cuda.use_mem_pool(pool):
    a = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
del a
# at the memory limit, this will succeed by using pool's memory in order to avoid the oom
b = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
```

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 commented:
@pytorchbot merge

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a use_on_oom bool with each MemPool.

Usage:
Users can optionally specify a ``use_on_oom`` bool (which is False by default) during MemPool creation. If true, then the CUDACachingAllocator will be able to use memory in this pool as a last resort instead of OOMing.

```
pool = torch.cuda.MemPool(allocator, use_on_oom=True)
with torch.cuda.use_mem_pool(pool):
    a = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
del a
# at the memory limit, this will succeed by using pool's memory in order to avoid the oom
b = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
```

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 commented:
@pytorchbot merge

dsjohns2 committed:
Update on "Add option to use mempool on OOM"

MemPool is a separate pool of memory handled by the caching allocator. This PR adds the option let the caching allocator try to use this pool as a last resort instead of OOMing by associating a use_on_oom bool with each MemPool.

Usage:
Users can optionally specify a ``use_on_oom`` bool (which is False by default) during MemPool creation. If true, then the CUDACachingAllocator will be able to use memory in this pool as a last resort instead of OOMing.

```
pool = torch.cuda.MemPool(allocator, use_on_oom=True)
with torch.cuda.use_mem_pool(pool):
    a = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
del a
# at the memory limit, this will succeed by using pool's memory in order to avoid the oom
b = torch.randn(40 * 1024 * 1024, dtype=torch.uint8, device="cuda")
```

Testing: 
```
python test/test_cuda.py -k test_mempool_limited_memory_with_allocator
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dsjohns2 commented:
@pytorchbot merge
